src/reference/03-Developers-Guide/02-Coding-Guideline/00.md

```diff
@@ -0,0 +1,204 @@
+
+Making independent modules with fewer dependent libraries make it easier to reuse them.
+
+#### Hide external classes
+
+Avoid exposing external classes out to API, except for standard Scala and Java classes.
+
+#### Hide internal modules
+
+A module may be declared internal if there's no use for public.
+
+#### Compiler flags
+
+```
+-encoding utf8
+-deprecation
+-feature
+-unchecked
+-Xlint
+-language:higherKinds
+-language:implicitConversions
+-Xfuture
+-Yinline-warnings
+-Yno-adapted-args
+-Ywarn-dead-code
+-Ywarn-numeric-widen
+-Ywarn-value-discard
+-Xfatal-warnings
+```
+
+The `-Xfatal-warnings` may be removed if there are unavoidable warnings.
+
+#### Package name and organization name
+
+Use the package name appended with the layer name, such as `sbt.io` for IO layer.
+The organization name for the publish artifacts should remain `org.scala-sbt`.
+
+### Binary resiliency
+
+A good overview on the topic of binary resiliency is [Josh's 2012][jsuereth2012] talk
+Binary resiliency.
+The guideline here applies mostly to publicly exposed APIs.
+
+#### MiMa
+
+Use [MiMa][mima].
+
+#### Public traits should contain `def` declarations only
+
+- `val` or `var` in `trait` results to codegen at subclass and at the artificial `Foo\$class.\$init\$`.
+- `lazy val` results to codegen at subclass
+
+#### Abstract classes are also useful
+
+[To trait, or not to trait?][pins_trait127].
+It is less flexible compared to trait, but easier to maintain binary compatibility. It also has better Java interop.
+
+#### Seal the traits and abstract classes
+
+If there's no need to keep the classes open, seal it.
+
+#### Finalize the leaf classes
+
+If there's no need to keep the classes open, finalize it.
+
+#### Typeclass and subclass inheritance
+
+Typeclass pattern with pure trait might be easier to maintain binary compatibility than subclassing.
+
+#### Avoid case classes, use sbt-datatype
+
+case class involves various codegen, which makes it harder to maintain binary compatibility over time.
+
+#### Prefer method overloading over default parameter values
+
+The default parameter values are effectively codegen,
+which makes it difficult to maintain.
+
+### Other public API matter
+
+Here are other guidelines about public API.
+
+#### Avoid Stringly-typed programming
+
+Define datatypes.
+
+#### Avoid overuse of `def apply`
+
+`def apply` should be reserved for factory methods
+in a companion object that returns type `T`.
+
+#### Use specific datatype (`Vector`, `List`, or `Array`) rather than `Seq`.
+
+`scala.Seq` is `scala.collection.Seq`, which is not immutable.
+Default to `Vector`. Use `List` if constant prepending is needed.
+Use `Array` if Java interop is needed.
+Note using mutable collection is perfectly fine within the implementation.
+
+#### Avoid calling `toSeq` or anything side-effecty on `Set`
+
+`Set` is fine if you stick with set operations, like `contains` and `subsetOf`.
+More often than not, `toSeq` is called explicitly or implicitly,
```

Comment:
I bet we could make an abide rule to detect this and throw an exception any time a `Set` appears in an expression of type `Seq` (or you see a toSeq)